### PR TITLE
Set picking type on PO generated from sourcing with FA

### PR DESCRIPTION
**Pull request by yvaucher**
This bridge with onchanges on purchase_delivery_address
and make it consistant with purchase_requisition_delivery_address

### CODE REVIEW

**Comment by gurneyalex** [comment]
please add a couple of automated tests

**Comment by yvaucher** [comment]
@gurneyalex Tests added

File: framework_agreement_sourcing/tests/test_agreement_souce_line_to_po.py [hunk 1]
```diff
@@ -52,9 +52,80 @@ def setUp(self):
         self.other_source = next(x for x in self.source_lines
                                  if x.procurement_method == 'other')
 
-    def test_01_transform_source_to_agreement(self):
+    def test_01_transform_source_to_agreement_wh_dest(self):
```
Comment (foutoucour) [inline framework_agreement_sourcing/tests/test_agreement_souce_line_to_po.py:55]: I think good practices advices to have small tests with one scenrio/case by test method

Comment (lepistone) [inline framework_agreement_sourcing/tests/test_agreement_souce_line_to_po.py:55]: @foutoucour can you please elaborate more? 

IMO we can have both 
- "real" isolated unit tests (setup - run one method - assert on the result)
- integration/story-like tests

In the core there are only integration ones, but that doesn't mean we shouldn't write isolated ones. I personally am for isolation wherever possible, but I'd rather avoid that big debate between isolation and integration and accept both styles.

**Comment by lepistone** [comment]
There are two tests numbered 01. Each one is run in its own transaction, so I don't think we need to number them at all.

Otherwise :+1: thanks!

**Comment by lepistone** [comment]
@foutoucour do you think we can merge this one? If not, what do you suggest? Thanks!

**Comment by gurneyalex** [comment]
:+1:

File: framework_agreement_sourcing/model/logistic_requisition_source.py [hunk 2]
```diff
@@ -60,6 +60,25 @@ def _company(self, cr, uid, context):
...
+            ('code', '=', 'incoming')])
+
+        picking_type_id = False
+        if types:
+            picking_type_id = types[0].id
+        elif dest_address.customer:
+            # if destination is not for a warehouse address,
+            # we set dropshipping picking type
+            ref = 'stock_dropshipping.picking_type_dropship'
+            picking_type_id = self.env['ir.model.data'].xmlid_to_res_id(ref)
```
Comment (nbessi) [inline framework_agreement_sourcing/model/logistic_requisition_source.py:79]: We might want to use env.ref and return a record directly?

File: framework_agreement_sourcing/model/logistic_requisition_source.py [hunk 3]
```diff
@@ -83,6 +104,8 @@ def _prepare_purchase_order(self, cr, uid, line, po_pricelist,
         data['company_id'] = self._company(cr, uid, context)
         data['pricelist_id'] = po_pricelist.id
         data['dest_address_id'] = add.id
+        if pick_type_id:
```
Comment (nbessi) [inline framework_agreement_sourcing/model/logistic_requisition_source.py:107]: Ok I get the point of the ID return :)

File: framework_agreement_sourcing/tests/test_agreement_souce_line_to_po.py (hunk 1 above)
Comment (nbessi) [inline framework_agreement_sourcing/tests/test_agreement_souce_line_to_po.py:55]: If there is parts we can reuse we should at lest split the block. It will ease redability

Comment (lepistone) [inline framework_agreement_sourcing/tests/test_agreement_souce_line_to_po.py:55]: @nbessi indeed nesting helper methods helps to DRY, but sometimes I personally find it hard to follow and somewhat verbose (less repetition but more boilerplate).

Instead often I prefer to have shorter, more expressive test by splitting them in functionally significant parts (small realistic scenarios instead of pages-long stories), and avoiding data irrelevant to the test (sometimes with help from `Model.new()` and mocks). A bit of functionally relevant repetition in an expressive, short test is fine by me.

This is 100% subjective preference, of course: we all just want to make our tests readable and maintainable.

Comment (foutoucour) [inline framework_agreement_sourcing/tests/test_agreement_souce_line_to_po.py:55]: I meant if it is a unittest I prefer to have one test by assert. The tests might be long at the end of the day but it is really helpful during debug I found.

I didn't realised that was not the case here and I understand why the test has so many  asserts.

**Comment by foutoucour** [comment]
:+1:
